src/Bicep.Core/Emit/ParameterAssignmentEvaluator.cs

```diff
@@ -408,4 +424,47 @@ private static ResultWithDiagnosticBuilder<Template> GetTemplate(SemanticModel m
             return new(x => x.ReferencedModuleHasErrors());
         }
     }
+
+
+    /// <summary>
+    /// Rewrites the external input function calls to use the externalInputs function with the index of the external input.
+    /// e.g. externalInput('sys.cli', 'foo') becomes externalInputs('0')
+    /// </summary>
+    private class ExternalInputExpressionRewriter : ExpressionRewriteVisitor
+    {
+        private readonly ExternalInputReferences externalInputReferences;
+
+        private ExternalInputExpressionRewriter(
+            ExternalInputReferences externalInputReferences)
+        {
+            this.externalInputReferences = externalInputReferences;
+        }
+
+
+        public static Expression Rewrite(
+            Expression expression,
+            ExternalInputReferences externalInputReferences)
+        {
+            var visitor = new ExternalInputExpressionRewriter(externalInputReferences);
+            var rewritten = visitor.Replace(expression);
+
+            return rewritten;
+        }
+
+        public override Expression ReplaceFunctionCallExpression(FunctionCallExpression expression)
+        {
+            if (string.Equals(expression.Name, LanguageConstants.ExternalInputArmFunctionName, LanguageConstants.IdentifierComparison) && 
```

Comment:
[nit] very much personal preference, but I find the following slightly more readable:

```csharp
LanguageConstants.IdentifierComparer.Equals(expression.Name, LanguageConstants.ExternalInputArmFunctionName)
```